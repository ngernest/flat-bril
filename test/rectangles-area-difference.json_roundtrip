{
  "functions": [
    {
      "args": [
        {
          "name": "x1",
          "type": "int"
        },
        {
          "name": "y1",
          "type": "int"
        },
        {
          "name": "x2",
          "type": "int"
        },
        {
          "name": "y2",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x1",
            "y1"
          ],
          "dest": "a1",
          "funcs": [
            "area"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x2",
            "y2"
          ],
          "dest": "a2",
          "funcs": [
            "area"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a1",
            "a2"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "a1",
            "a2"
          ],
          "dest": "a1_bigger",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "a1_bigger"
          ],
          "funcs": [],
          "labels": [
            "end",
            "flip"
          ],
          "op": "br"
        },
        {
          "label": "flip"
        },
        {
          "dest": "neg1",
          "op": "const",
          "type": "int",
          "value": -1
        },
        {
          "args": [
            "res",
            "neg1"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "res"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "area",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "area"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "area"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "area",
      "type": "int"
    }
  ]
}
