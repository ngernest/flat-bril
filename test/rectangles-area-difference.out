@main
	op: call 		dest: a1		type: int		args: ["x1", "y1"]		funcs: ["area"]
	op: call 		dest: a2		type: int		args: ["x2", "y2"]		funcs: ["area"]
	op: sub  		dest: res		type: int		args: ["a1", "a2"]	
	op: gt   		dest: a1_bigger		type: bool		args: ["a1", "a2"]	
	op: br   		args: ["a1_bigger"]		labels: ["end", "flip"]
.flip
	op: const		dest: neg1		type: int		value: -1	
	op: mul  		dest: res		type: int		args: ["res", "neg1"]	
.end
	op: print		args: ["res"]	
@area
	op: mul  		dest: area		type: int		args: ["x", "y"]	
	op: print		args: ["area"]	
	op: ret  		args: ["area"]	
